### ISSUE TITLE
Languages branch and setup languages

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#981]
### Expected behaviour

Because we now have languages as flat files in the languages branch, we should be pulling the setup language strings from those files **not** from a separate file `setup-langs.php` which doubles the effort required to maintain language Textpacks. I'm sure there must be an efficient way of pulling all strings from `@setup` group for your chosen language at setup?

### Actual behaviour

Currently strings for setup read `setup-langs.php` file, which has a mirror of strings already in each Textpack.

**Comment by jools-r** [comment]
Having just looked at this for a revised multi-site setup, I agree. One question, though: can `{variables}` be used in the textpack lang files in the same way they are in setup-langs.php to inject config/setup path details into the language string?

**Comment by philwareham** [comment]
@bloke thoughts? I don't know but I guess yes!

**Comment by Bloke** [comment]
The `setup_gTxt()` function used in setup differs from the regular one in an important way: under regular usage, `gTxt()` returns the raw token if it doesn't match. Setup detects if nothing matched and returns the translated _English_ string as fallback. Other than that, they are identical.

This poses a slight problem for fallback languages as we have to read two separate files during setup: chosen language and English, just in case, but we could probably be clever here:

1. read both files into separate arrays.
2. merge the arrays but use the English strings to fill in holes. Array concatenation does that by default in PHP.

That gives us one massive replacement block with all strings. We can then remove any in the 'setup' group that we have elsewhere so the setup group only contains unique ones to the setup process.

In short, I don't see any impediment to using the installed languages now as long as:

* We read the files directly and don't use `load_lang()` because the database doesn't exist!
* We definitely, definitely omit strings from the 'setup' group when they are installed into the DB. I think that `L10n/Lang.php` already does this (see the `install_file` method) but I've not checked if it actually works.

**Comment by philwareham** [comment]
Just this one left to do I think, and then the languages branch can be merged into dev! \o/

**Comment by Bloke** [comment]
On it...

**Comment by Bloke** [comment]
See what you make of c81645a8. Seems to have improved matters, but please test installing textpack strings too after installation.

Also fixes a bug during setup where the language was lost if you went back a step if you entered incorrect credentials in step 1.

**Comment by philwareham** [comment]
Sorry, a couple of minor things I spotted.

1. The pophelp dialogs title 'help' is empty. It's to do with this line 2168 in our textpattern.js file:

    title: textpattern.gTxt('help'),

That string is outside of setup group. Not sure if we can simply duplicate that string in setup or if it needs to be a unique (e.g. like we have namespaced `setup_show_password => Show password`). If so it'll need some extra conditional logic in the JS file to check where this is a setup page or not, and supply...

     title: textpattern.gTxt('setup_help'),

...instead?

2. An edge case but worth mentioning:

If you select a language and proceed to the 'set database details' page.
Then use back button in your browser to return to initial 'Select language' page.
Then select a different language and click continue (to return to 'set database details' page).

It then doesn't use the new language choice, keeps the original language choice.

**Comment by Bloke** [comment]
1, Does it help if 'help' is oput in the common group? Or do we need to explicitly add it to the `gTxtScript` array?
2. Ah nuts, that was my meddilng with the back button thingy trying to get it to remember the content that's already set. I'll fiddle with that, thanks for testing.

**Comment by philwareham** [comment]
I'll test. IMO `help =>` should be moved to common anyway. Will get back to you (probably tomorrow now - got deadlines today).

**Comment by philwareham** [comment]
Hi again @Bloke - sorry it makes no difference having `help =>` in `common` group when it comes to setup (although `common` is right place for that particular string). Still outputs the raw string name.

I think you'll need to do some coding magic - and then I can make a new string in setup group called `setup_help =>`.

**Comment by Bloke** [comment]
Nah, I think I know what's going on. Let's put `help` in the `common` group if it's not there already. Then I'll just need to add it to the list of strings that get added to `gTxtScript` on the page so that JavaScript can access them.

Note to self: must be a better way to do this overall.

**Comment by philwareham** [comment]
Ok it’s in common group.

**Comment by Bloke** [comment]
Both issues should now be fixed. Please retest.

**Comment by philwareham** [comment]
@bloke Yep all working fine now. Many thanks. Closing this one.
